server/STYLE.md

```diff
@@ -0,0 +1,505 @@
+# Code conventions and style guide
+
+This is a short document describing the preferred coding style for this project. We've tried to cover the major areas of formatting and naming. When something isn't covered by this guide, err on the side of consistency with existing code.
+
+### Formatting
+
+Use [stylish-haskell](https://github.com/jaspervdj/stylish-haskell) to format your code.
+
+#### Line Length
+
+Try to keep a line length within 100 characters. Small exceptions are okay if the line only goes over the limit by a few characters and breaking it would negatively impact readability. Use your best judgement.
```

Comment:
The problem I have with 100 characters is that it doesn't work well in settings where you want to open an another window on the side (say hoogle) or view more than one file at the same time. On my 15.6" laptop with 12pt font, this is how it looks:

A single file, looks great! 

![image](https://user-images.githubusercontent.com/6562944/78207532-983e0b80-74bf-11ea-9ca9-d4a2e64d3d21.png)

Now if I have two files open:

![image](https://user-images.githubusercontent.com/6562944/78207992-e0a9f900-74c0-11ea-9e3a-2f8ffc24cfbe.png)

This is how it looks with a max line-length of 85 characters (formatted with vim's `gw`)

![image](https://user-images.githubusercontent.com/6562944/78208380-e3591e00-74c1-11ea-9fee-3d69c2e01811.png)

The readability is much better with line-length at at 85 characters in this case. Now the question is how common is this pattern of having something else next to the file that you are editing, be it another file or a browser window, I honestly don't know, but I think we should be accommodating to devs who prefer such a workflow if it isn't too much of an inconvenience to others. I could be wrong but on a 14" laptop with 11pt font, 80-85 character limit should fit two files without any line wrapping.

Reply:
For what it’s worth, I’ve started feeling like wrapping *comments* at 100 characters is a bad idea, too. I rarely do horizontal splits, but even without them, the wide lines are just less pleasant to read! More recently, I’ve been wrapping all my comments at 70–80 characters rather than 100.

That said, I really don’t like an 80 character limit on code. I would much rather see

```haskell
      loggerCtx <- liftIO $ mkLoggerCtx (defaultLoggerSettings True logLevel) enabledLogs
```

than

```haskell
      loggerCtx <- liftIO $ mkLoggerCtx (defaultLoggerSettings True logLevel)
                                        enabledLogs
```

just because the latter was slightly too long. In my experience, setting a limit at 80 characters encourages two bad habits:

  1. It makes people less willing to create more nested binding contexts. Suddenly putting a `where` inside a `where` has a cost—instead of 80 characters, you get just 72. Now have a `let` inside a `do` block in that nested binding, and you’re working with 66 characters. Things get cramped quickly, so people tend to float things out more… but I think that often hurts readability more than it helps!

  2. It encourages shorter names. In the example I gave above, I might question whether I needed to use all those characters. Could I use `defLogSettings` and `logLvl`? Might as well write `mkPermMetaObj` instead of `mkPermissionMetadataObject`… or something even more cryptic.

     I would much rather have to soft wrap or horizontally scroll a little bit and have nice, readable variable names than stare at abbreviations that don’t make sense to me!

So I’d be happy to tighten up the length guideline for comments, but I’d rather not go below 100 on the code line length limit. (If anything, be glad I’m not lobbying for something even *longer.* 😁)

Reply:
When we use stop using cryptic variable names, I think we'll see the line length go beyond 100 characters more often than not (even in the above example, expanding `Ctx` to `Context` will result in a much longer line), so I think that we will need to get used to such things:

```haskell
      loggerCtx <- liftIO $ mkLoggerCtx (defaultLoggerSettings True logLevel)
                                        enabledLogs
```

In terms of readability, I prefer this to soft-wrap any day because soft wraps don't respect indentation! Horizontal scroll is equally bad I feel, you can't see everything with a quick glance.

> It makes people less willing to create more nested binding contexts. Suddenly putting a where inside a where has a cost—instead of 80 characters, you get just 72. Now have a let inside a do block in that nested binding, and you’re working with 66 characters. Things get cramped quickly, so people tend to float things out more… but I think that often hurts readability more than it helps!

I agree with all this but at the same time I think 100 characters too won't be enough with our naming convention! How about a soft limit at 85-90 characters but if the developer thinks that using more characters improves readability they can definitely do that?

Reply:
I have updated the section on line length to suggest a target length of 80 characters. Let me know if you’re happy with the new wording!